From 406fb1ba3ff84fd589e97104dd91b58376721402 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Wed, 18 Dec 2024 00:25:54 +0100
Subject: [PATCH 42/86] win32u/tests: debugstr_w - debugstr_wn.

/home/bernhard/data/entwicklung/2024/wine/wine/tools/runtest -q -P wine -T . -M win32u.dll -p dlls/win32u/tests/i386-windows/win32u_test.exe win32u && touch dlls/win32u/tests/i386-windows/win32u.ok
=================================================================
==2640==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x005cf650 at pc 0x0042d112 bp 0x005cdf34 sp 0x005cdf30
READ of size 2 at 0x005cf650 thread T0
0a64:fixme:msvcrt:_set_abort_behavior _WRITE_CALL_REPORTFAULT unhandled
0a64:fixme:file:server_get_file_info Unsupported info class e
    #0 0x0042d111 in wine_dbgstr_wn /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/debug.h:263:30
    #1 0x0040394a in debugstr_w /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/debug.h:522:65
    #2 0x0040394a in test_class /home/bernhard/data/entwicklung/2024/wine\wine/dlls/win32u/tests/win32u.c:269:41
    #3 0x0040394a in func_win32u /home/bernhard/data/entwicklung/2024/wine\wine/dlls/win32u/tests/win32u.c:2548:5
    #4 0x00436849 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #5 0x00436260 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h
    #6 0x00437f0b in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #7 0x7877ea3f in BaseThreadInitThunk (C:\windows\system32\kernel32.dll+0x7b80ea3f)
    #8 0x788fce82 in call_thread_func_wrapper (C:\windows\system32\ntdll.dll+0x7bc0ce82)
    #9 0x789304b4 in call_thread_func /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\signal_i386.c:524:9

Address 0x005cf650 is located in stack of thread T0 at offset 5296 in frame
    #0 0x0040100f in func_win32u /home/bernhard/data/entwicklung/2024/wine\wine/dlls/win32u/tests/win32u.c:2518

  This frame has 37 object(s):
    [16, 100) 'source_name.i' (line 1291)
    [144, 184) 'cls.i675' (line 2031)
    [224, 264) 'cls.i638' (line 1349)
    [304, 564) 'path.i' (line 1350)
    [640, 656) 'pi.i' (line 1351)
    [672, 740) 'startup.i' (line 1352)
    [784, 812) 'msg.i639' (line 1353)
    [848, 876) 'msg.i615' (line 1271)
    [912, 940) 'msg.i' (line 1220)
    [976, 1024) 'item.i' (line 1143)
    [1056, 2080) 'buf.i537' (line 1088)
    [2208, 2216) 'callback_params.i' (line 1024)
    [2240, 2252) 'smp.i' (line 1028)
    [2272, 2312) 'cls.i440' (line 1029)
    [2352, 2872) 'module.i' (line 891)
    [3008, 3528) 'res_buf.i' (line 891)
    [3664, 3672) 'module_str.i' (line 892)
    [3696, 3704) 'res_str.i' (line 892)
    [3728, 4752) 'bmp_bits.i' (line 893)
    [4880, 4884) 'width.i' (line 894)
    [4896, 4900) 'height.i' (line 894)
    [4912, 4916) 'rate.i' (line 895)
    [4928, 4932) 'steps.i' (line 895)
    [4944, 4964) 'info.i348' (line 898)
    [5008, 5012) 'ret_size.i' (line 752)
    [5024, 5056) 'buf.i234' (line 415)
    [5088, 5100) 'himc.i' (line 415)
    [5120, 5124) 'size.i' (line 417)
    [5136, 5144) 'name.i' (line 180)
    [5168, 5296) 'buf.i173' (line 181) <== Memory access at offset 5296 overflows this variable
    [5328, 5368) 'cls.i' (line 182)
    [5408, 5412) 'count.i' (line 123)
    [5424, 6264) 'info.i' (line 71)
    [6400, 6436) 'mdi.i' (line 1389)
    [6480, 6680) 'bufW.i' (line 1390)
    [6752, 6852) 'buf.i' (line 1391)
    [6896, 6900) 'argv' (line 2519)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/debug.h:263:30 in wine_dbgstr_wn
Shadow bytes around the buggy address:
  0x005cf380: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x005cf400: f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2
  0x005cf480: f2 f2 f2 f2 f2 f2 f8 f2 f8 f2 f8 f2 f8 f2 f8 f8
  0x005cf500: f8 f2 f2 f2 f2 f2 f8 f2 f8 f8 f8 f8 f2 f2 f2 f2
  0x005cf580: f8 f8 f2 f2 f8 f2 00 f2 f2 f2 00 00 00 00 00 00
=>0x005cf600: 00 00 00 00 00 00 00 00 00 00[f2]f2 f2 f2 00 00
  0x005cf680: 00 00 00 f2 f2 f2 f2 f2 f8 f2 f8 f8 f8 f8 f8 f8
  0x005cf700: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x005cf780: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x005cf800: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x005cf880: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==2640==ABORTING
make: *** [Makefile:414149: dlls/win32u/tests/i386-windows/win32u.ok] Fehler 1
---
 dlls/win32u/tests/win32u.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/win32u/tests/win32u.c b/dlls/win32u/tests/win32u.c
index b51218ef60d..2db0572a3eb 100644
--- a/dlls/win32u/tests/win32u.c
+++ b/dlls/win32u/tests/win32u.c
@@ -266,7 +266,7 @@ static void test_class(void)
     ret = NtUserGetAtomName( class, &name );
     ok( !ret && GetLastError() == ERROR_INVALID_HANDLE,
         "NtUserGetAtomName returned %lx %lu\n", ret, GetLastError() );
-    ok( buf[0] == 0xcccc, "buf = %s\n", debugstr_w(buf) );
+    ok( buf[0] == 0xcccc, "buf = %s\n", debugstr_wn(buf, ARRAY_SIZE(buf)) );
 
 }
 
-- 
2.47.1

